### PR TITLE
New: Simkl List Support

### LINKED ISSUE #5306: Simkl List integration
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

In light of the current issues Trakt are having with their database it may be prudent to add other list integrations to sonarr such as simkl etc so that people that use both would be able to add one as a back up incase the other goes down

### Describe the solution you'd like

Simkl Lists as an option in import lists

### Describe alternatives you've considered

N/A

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Qstick**
#### Database Migration
NO

#### Description
Adds support for Simkl User Watchlists (Watched/PlanToWatch/Hold/Completed/Dropped). OAuth app has been setup on https://auth.servarr.com similar to Trakt. 

#### Todos
- [ ] Wiki Updates

#### Issues Fixed or Closed by this PR

* Fixes #5306

### CODE REVIEW

File: src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            var series = new List<ImportListItemInfo>();
+
+            if (!PreProcess(_importResponse))
+            {
+                return series;
+            }
+
+            var jsonResponse = JsonConvert.DeserializeObject<SimklResponse>(_importResponse.Content);
+
+            // no movies were return
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs:31]: `No series were returned`

File: src/NzbDrone.Core/ImportLists/Simkl/User/SimklUserRequestGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                    break;
+                case (int)SimklUserListType.Completed:
+                    link += $"/sync/all-items/shows/completed";
+                    break;
+                case (int)SimklUserListType.Dropped:
+                    link += $"/sync/all-items/shows/dropped";
+                    break;
+            }
+
+            var request = new ImportListRequest($"{link}", HttpAccept.Json);
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/User/SimklUserRequestGenerator.cs:49]: Link is already a string, no need to interpolate it

File: src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            try
+            {
+                var response = _httpClient.Get<RefreshRequestResponse>(request);
+
+                if (response != null && response.Resource != null)
+                {
+                    var token = response.Resource;
+                    Settings.AccessToken = token.AccessToken;
+                    Settings.Expires = DateTime.UtcNow.AddSeconds(token.ExpiresIn);
+                    Settings.RefreshToken = token.RefreshToken != null ? token.RefreshToken : Settings.RefreshToken;
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs:168]: Could simplify this to:

```
Settings.RefreshToken = token.RefreshToken ?? Settings.RefreshToken;
```

File: src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            var lastActivity = GetLastActivity();
+
+            // Check to see if user has any activity since last sync, if not return empty to avoid work
+            if (lastActivity < lastFetch.AddHours(-2))
+            {
+                return new List<ImportListItemInfo>();
+            }
+
+            var generator = GetRequestGenerator();
+            return FetchItems(g => g.GetListItems(), true);
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs:58]: Nit: newline before return

**Review by markus101 — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs (hunk 1 above)
Comment (Qstick) [reply src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs:31]: Don't go look at the Trakt implementation :D

File: src/NzbDrone.Core/ImportLists/Simkl/SimklSettingsBase.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            BaseUrl = "https://api.simkl.com";
+            SignIn = "startOAuth";
+        }
+
+        public string BaseUrl { get; set; }
+
+        [FieldDefinition(0, Label = "Access Token", Type = FieldType.Textbox, Hidden = HiddenType.Hidden)]
+        public string AccessToken { get; set; }
+
+        [FieldDefinition(0, Label = "Refresh Token", Type = FieldType.Textbox, Hidden = HiddenType.Hidden)]
```
[the remark this answers is not in the document]
Comment (Qstick) [reply src/NzbDrone.Core/ImportLists/Simkl/SimklSettingsBase.cs:43]: All of these are hidden which was the reason for all 0s (similar to Trakt). Should we still number them?

[the remark this answers is not in the document]
Comment (markus101) [reply src/NzbDrone.Core/ImportLists/Simkl/SimklSettingsBase.cs:43]: ahh, duh, no I don't think so then

**Comment by Qstick** [comment]
I think that should address the comments, also fixed a slight bug with the activity check and switched the de-serialize to STJson from Newtonsoft

File: src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            var request = new HttpRequestBuilder(string.Format("{0}/sync/activities", Settings.BaseUrl)).Build();
+
+            request.Headers.Add("simkl-api-key", ClientId);
+            request.Headers.Add("Authorization", "Bearer " + Settings.AccessToken);
+
+            try
+            {
+                var response = _httpClient.Get<SimklSyncActivityResource>(request);
+
+                if (response != null && response.Resource != null)
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs:108]: Could shorten this to `response?.Resource != null`, probably worth getting into that habbit.

File: src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            if (accessToken.IsNotNullOrWhiteSpace())
+            {
+                request.Headers.Add("Authorization", "Bearer " + accessToken);
+            }
+
+            try
+            {
+                var response = _httpClient.Get<UserSettingsResponse>(request);
+
+                if (response != null && response.Resource != null)
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs:137]: Same thing here: `response?.Resource != null`

File: src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+            var request = new HttpRequestBuilder(RenewUri)
+                .AddQueryParam("refresh_token", Settings.RefreshToken)
+                .Build();
+
+            try
+            {
+                var response = _httpClient.Get<RefreshRequestResponse>(request);
+
+                if (response != null && response.Resource != null)
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/SimklImportBase.cs:164]: Same thing here: `response?.Resource != null`

File: src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        protected virtual bool PreProcess(ImportListResponse netImportResponse)
+        {
+            if (netImportResponse.HttpResponse.StatusCode != HttpStatusCode.OK)
+            {
+                throw new ImportListException(netImportResponse, "Simkl API call resulted in an unexpected StatusCode [{0}]", netImportResponse.HttpResponse.StatusCode);
+            }
+
+            if (netImportResponse.HttpResponse.Headers.ContentType != null && netImportResponse.HttpResponse.Headers.ContentType.Contains("text/json") &&
```
Comment (markus101) [inline src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs:57]: Could also do something like:

```
if ((netImportResponse.HttpResponse.Headers.ContentType?.Contains("text/json") ?? false) &&
                (!netImportResponse.HttpRequest.Headers.Accept?.Contains("text/json") ?? false))
```

But I think it's a lot harder to follow and more fragile.

**Review by markus101 — APPROVED** [review]
Just a few comments on using a null conditional and one where we probably shouldn't, but looks good.

File: src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs (hunk 9 above)
Comment (Qstick) [reply src/NzbDrone.Core/ImportLists/Simkl/SimklParser.cs:57]: Yea, agree seems a bit less readable, and this is consistent with other examples in the code base.
